docs/development/core/public/kibana-plugin-public.contextsetup.md

```diff
@@ -0,0 +1,132 @@
+
+## Methods
+
+|  Method | Description |
+|  --- | --- |
+|  [createContextContainer()](./kibana-plugin-public.contextsetup.createcontextcontainer.md) | Creates a new [ContextContainer](./kibana-plugin-public.contextcontainer.md) for a service owner. |
+
+## Remarks
+
+A [ContextContainer](./kibana-plugin-public.contextcontainer.md) can be used by any Core service or plugin (known as the "service owner") which wishes to expose APIs in a handler function. The container object will manage registering context providers and configuring a handler with all of the contexts that should be exposed to the handler's plugin. This is dependent on the dependencies that the handler's plugin declares.
+
+Contexts providers are executed in the order they were registered. Each provider gets access to context values provided by any plugins that it depends on.
+
+In order to configure a handler with context, you must call the [ContextContainer.createHandler()](./kibana-plugin-public.contextcontainer.createhandler.md) function and use the returned handler which will automatically build a context object when called.
+
+When registering context or creating handlers, the \_calling plugin's id\_ must be provided. Note this should NOT be the context service owner, but the plugin that is actually registering the context or handler.
+
+```ts
+// GOOD
+class MyPlugin {
+  private readonly handlers = new Map();
+
+  setup(core) {
+    this.contextContainer = core.context.createContextContainer();
+    return {
+      registerContext(plugin, contextName, provider) {
+        this.contextContainer.registerContext(plugin, contextName, provider);
+      },
+      registerRoute(plugin, path, handler) {
+        this.handlers.set(
+          path,
+          this.contextContainer.createHandler(plugin, handler)
+        );
+      }
+    }
+  }
+}
+
+// BAD
+class MyPlugin {
+  private readonly handlers = new Map();
+
+  setup(core) {
+    this.contextContainer = core.context.createContextContainer();
+    return {
+      registerContext(plugin, contextName, provider) {
+        // This would leak this context to all handlers rather tha only plugins that depend on the calling plugin.
+        this.contextContainer.registerContext('my_plugin', contextName, provider);
+      },
+      registerRoute(plugin, path, handler) {
+        this.handlers.set(
+          path,
+          // the handler will not receive any contexts provided by other dependencies of the calling plugin.
+          this.contextContainer.createHandler('my_plugin', handler)
+        );
+      }
+    }
+  }
+}
+
+```
+
+## Example
+
+Say we're creating a plugin for rendering visualizations that allows new rendering methods to be registered. If we want to offer context to these rendering methods, we can leverage the ContextService to manage these contexts.
+
+```ts
+export interface VizRenderContext {
+  core: {
+    i18n: I18nStart;
+    uiSettings: UISettingsClientContract;
+  }
+  [contextName: string]: unknown;

```

Comment:
is it a necessary declaration?

Reply:
Maybe not. I thought it could be useful to help show that the `contextName` parameter in `registerContext` corresponds to keys on this object.